org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JdtSourceLookUpProvider.java

```diff
@@ -0,0 +1,139 @@
+     */
+    private Object findSourceElement(String sourcePath, ISourceContainer[] containers) {
+        if (containers == null) {
+            return null;
+        }
+        for (ISourceContainer container : containers) {
+            try {
+                Object[] objects = container.findSourceElements(sourcePath);
+                if (objects.length > 0 && (objects[0] instanceof IResource || objects[0] instanceof IClassFile)) {
+                    return objects[0];
+                }
+            } catch (CoreException e) {
+                JavaLanguageServerPlugin.logException("Failed to find the source elements", e);
+            }
+        }
+        return null;
+    }
+
+    private ISourceContainer[] getSourceContainers(List<String> projectNames) {
+        List<IProject> projects = new ArrayList<>();
+        if (projectNames == null || projectNames.size() == 0) {
+            projects.addAll(Arrays.asList(ProjectUtils.getAllProjects()));
+        } else {
+            for (String projectName : projectNames) {
+                projects.add(ProjectUtils.getProject(projectName));
+            }
+        }
+
+        Set<ISourceContainer> containers = new LinkedHashSet<>();
+        Set<IRuntimeClasspathEntry> calculated = new LinkedHashSet<>();
+        projects.stream().map(project -> ProjectUtils.getJavaProject(project))
+            .filter(javaProject -> javaProject != null && javaProject.exists())
+            .forEach(javaProject -> {
+                // Add source containers associated with the project's runtime classpath entries.
+                containers.addAll(Arrays.asList(getSourceContainers(javaProject, calculated)));
+                // Add source containers associated with the project's source folders.
+                containers.add(new JavaProjectSourceContainer(javaProject));
+            });
+
+        return containers.toArray(new ISourceContainer[0]);
+    }
+
+    private ISourceContainer[] getSourceContainers(IJavaProject project, Set<IRuntimeClasspathEntry> calculated) {
+        if (project == null || !project.exists()) {
+            return new ISourceContainer[0];
+        }
+
+        try {
+            IRuntimeClasspathEntry[] unresolved = JavaRuntime.computeUnresolvedRuntimeClasspath(project);
+            List<IRuntimeClasspathEntry> resolved = new ArrayList<>();
+            for (IRuntimeClasspathEntry entry : unresolved) {
+                for (IRuntimeClasspathEntry resolvedEntry : JavaRuntime.resolveRuntimeClasspathEntry(entry, project)) {
+                    if (!calculated.contains(resolvedEntry)) {
+                        calculated.add(resolvedEntry);
+                        resolved.add(resolvedEntry);
+                    }
+                }
+            }
+            Set<ISourceContainer> containers = new LinkedHashSet<>();
+            containers.addAll(Arrays.asList(
+                    JavaRuntime.getSourceContainers(resolved.toArray(new IRuntimeClasspathEntry[0]))));
```

Comment:
I was noticing that when `/home/user/.m2/repository/junit/junit/4.13/junit-4.13.jar[CPE_LIBRARY][K_BINARY][sourcePath:/home/user/.m2/repository/junit/junit/4.13/junit-4.13-sources.jar][isExported:false][attributes:maven.pomderived=true,test=true,maven.groupId=junit,maven.artifactId=junit,maven.version=4.13,maven.scope=test]` gets passed in, the source path never gets discovered.